### PR TITLE
update to support binary files instead of absolute path uploads

### PR DESCRIPTION
**Pull request by mdemin914**
Simple change to support binaryFile input instead of having to use fs.createReadStream and have the file saved locally.

This is mainly to support the use case of a user uploading a file to a form and I would like to upload it to Zendesk without having to save it to disk. In the current functionality, you had to specify an absolute path for fs.createReadStream to pick up the file. This is not ideal as I was not storing the uploaded files anywhere.

I got this to work without changes, by saving the file to disk before calling the upload method, but then ran into a 'write after end' error after the first successful file upload. If I didn't restart the server it would always fail the second time the upload call was made. It seems to be similar to [162](https://github.com/blakmatrix/node-zendesk/issues/162) and [167](https://github.com/blakmatrix/node-zendesk/issues/167)

Removing the use of fs seems to have resolved the 'write after end' issue as well as allowing me to directly upload files without having to save them.

### CODE REVIEW

**Comment by tmaclean-LV** [comment]
This is very helpful! Any chance this can be reopened and included in the main repo?

**Comment by rlemaire** [comment]
Great ! Thanks a lot !

I want this in the main repo too.

**Comment by blakmatrix** [comment]
affects https://github.com/blakmatrix/node-zendesk/pull/190

**Comment by mdemin914** [comment]
is there something I can do?

File: lib/client/client.js [hunk 1]
```diff
@@ -185,13 +186,20 @@ Client.prototype.requestUpload = function (uri, file, callback) {
...
+  self.emit('debug::request', self.options);
+  // assumes a binary file is passed in instead of a path
+  if(binary){
+    self.options.body = file;
+    return this._request(self.options, function (err, response, result) {
+      requestCallback(self, err, response, result, callback);
+    });
+  }
+  else{
+    let out = this._request(self.options, function (err, response, result) {
```
Comment (erikzrekz) [inline lib/client/client.js]: `let` would mean everyone needs to be at Node 4.X or above. Can we assume that yet?

**Comment by Germain67** [comment]
Hello, do you have any news on this pull request ?

I'm struggling with the same issue and it would really help if this pull request was accepted !

Thanks ;)

**Comment by CowanSM** [comment]
Trying this on 1.3.0 seems to fail with zendesk coming back with a 422:
{"error":"RecordInvalid","description":"Record validation errors","details":{"content_type":[{"description":"Content type cannot be blank","error":"BlankValue"}

**Comment by macrosak** [comment]
@blakmatrix can you merge this please?

**Pull request merged**